src/nvim/eval/typval_encode.c.h

```diff
@@ -0,0 +1,766 @@
+        cur_mpsv->data.l.li = cur_mpsv->data.l.li->li_next;
+        break;
+      }
+      case kMPConvPartial: {
+        partial_T *const pt = cur_mpsv->data.p.pt;
+        tv = cur_mpsv->tv;
+        switch (cur_mpsv->data.p.stage) {
+          case kMPConvPartialArgs: {
+            TYPVAL_ENCODE_CONV_FUNC_BEFORE_ARGS(tv,
+                                                pt == NULL ? 0 : pt->pt_argc);
+            cur_mpsv->data.p.stage = kMPConvPartialSelf;
+            if (pt != NULL && pt->pt_argc > 0) {
+              TYPVAL_ENCODE_CONV_LIST_START(NULL, pt->pt_argc);
+              _mp_push(mpstack, ((MPConvStackVal) {
+                .type = kMPConvPartialList,
+                .tv = NULL,
+                .data = {
+                  .a = {
+                    .arg = pt->pt_argv,
+                    .argv = pt->pt_argv,
+                    .todo = (size_t)pt->pt_argc,
+                  },
+                },
+              }));
+            }
+            break;
+          }
+          case kMPConvPartialSelf: {
+            cur_mpsv->data.p.stage = kMPConvPartialEnd;
+            dict_T *const dict = pt == NULL ? NULL : pt->pt_dict;
+            if (dict != NULL) {
+              TYPVAL_ENCODE_CONV_FUNC_BEFORE_SELF(tv, dict->dv_hashtab.ht_used);
+              const int te_csr_ret = _TYPVAL_ENCODE_CHECK_SELF_REFERENCE(
+                  TYPVAL_ENCODE_FIRST_ARG_NAME,
+                  dict, &dict->dv_copyID, &mpstack, copyID, kMPConvDict,
+                  objname);
+              if (te_csr_ret != NOTDONE) {
+                if (te_csr_ret == FAIL) {
+                  goto encode_vim_to__error_ret;
+                } else {
+                  continue;
+                }
+              }
+              TYPVAL_ENCODE_CONV_DICT_START(NULL, pt->pt_dict,
+                                            dict->dv_hashtab.ht_used);
+              _mp_push(mpstack, ((MPConvStackVal) {
+                .type = kMPConvDict,
+                .tv = NULL,
+                .data = {
+                  .d = {
+                    .dict = dict,
+                    .dictp = &pt->pt_dict,
+                    .hi = dict->dv_hashtab.ht_array,
+                    .todo = dict->dv_hashtab.ht_used,
+                  },
+                },
+              }));
+            } else {
+              TYPVAL_ENCODE_CONV_FUNC_BEFORE_SELF(tv, -1);
+            }
+            break;
+          }
+          case kMPConvPartialEnd: {
+            TYPVAL_ENCODE_CONV_FUNC_END(tv);
+            (void)_mp_pop(mpstack);
+            break;
+          }
+        }
+        continue;
+      }
+      case kMPConvPartialList: {
+        if (!cur_mpsv->data.a.todo) {
+          (void)_mp_pop(mpstack);
+          TYPVAL_ENCODE_CONV_LIST_END(NULL);
+          continue;
+        } else if (cur_mpsv->data.a.argv != cur_mpsv->data.a.arg) {
+          TYPVAL_ENCODE_CONV_LIST_BETWEEN_ITEMS(NULL);
+        }
+        tv = cur_mpsv->data.a.arg++;
```

Comment:
A quick check shows that first argument is freed:

    bin/nvim -u NONE -i NONE --cmd $'let obj={}\nfu obj.f()\nendfu' --cmd 'let Pt=function("string", [obj])|unlet obj|echo exists("*{1}")|unlet Pt|echo exists("*{1}")' --cmd cq

: if it is not freed then this will echo `1\n1`, otherwise `1\n0` which I see.